keps/sig-auth/20190408-node-scoped-daemonset.md

```diff
@@ -0,0 +1,378 @@
+
+## Definition
+
+**DaemonSet / Cluster-scoped DaemonSet**: a DaemonSet which allows a DaemonSet pod
+to manage the resources in a whole cluster.
+
+<!--
+**Node-scoped resources**: the resources covered by NodeAuthorizer’s [resource
+graph](https://github.com/kubernetes/kubernetes/blob/4db3fccd6f1c5ae8deb687985a2c8be76e4d4ab5/plugin/pkg/auth/authorizer/node/graph.go#L78),
+currently there are [8
+types](https://github.com/kubernetes/kubernetes/blob/d99f49d2a7caa2e86cb27541a267577771e133a7/plugin/pkg/auth/authorizer/node/graph.go#L123)
+of these resources: node, pod, configmap, pvc, pv, secret, serviceaccount, and
+volumeattachment.
+-->
+
+**Node-scoped resource access permission / node-scoped permission**: the
+permission to access a certain resource type is a subset of kubelet's permission
+on the resource type.
+
+**Node-scoped resource spec**: a string defining the resource types upon
+which a DaemonSet's permission should be node-scoped.
+
+**Node-scoped DaemonSet**: a DaemonSet includig a node-scoped resource spec.
+
+## Proposal
+
+Allow a DaemonSet owner to specify node-scoped resource spec (see the [Origin of
+Node-Scoped Resource Spec](#the-origin-of-node-scoped-resource-spec) section);
+APIServer tracks the node-scoped resource spec using a new string field
+(`NodeScopedResources string`) of
+[ServiceAccountInfo](https://github.com/kubernetes/kubernetes/blob/master/pkg/serviceaccount/util.go#L43),
+and adds the node-scoped resource spec into the
+[Extra](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L50)
+field of
+[user.DefaultInfo](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L46)
+in the
+[ServiceAccountInfo.UserInfo](https://github.com/kubernetes/kubernetes/blob/46c8b075ded386d99de1cf3945168127403c12e9/pkg/serviceaccount/util.go#L48)
+method during authentication. The node-scoped resource spec, together with the
+pod name/UID information in
+[user.DefaultInfo](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L46),
+can then be utilized in the authorization step and the admission step to implement
+node-scoped DaemonSet.
+
+  * `authorize a request on a node-scoped resource type`: when the type of the
+    resource accessed by the request is in the node-scoped resource spec, the
+    request will go through both NodeAuthorizer and RBAC (Note: the order
+    matters here), and the request must be explicitly allowed by the
+    NodeAuthorizer first before RBAC authorizes the request. If NodeAuthorizer
+    does not explicitly allow the request, the request will be denied directly.
+
+### The Origin of Node-Scoped Resource Spec
+
+Node-scoped info can be defined in two places:
+
+* Option 1: in
```

Comment:
I am still trying to grasp all of the pieces here, but adding a new field to SA seems wrong to me.  Could this be handled via the token request API?